Add more granular padding: top,right,bottom,left

sodiumjoe opened the pull request.

sodiumjoe commented on src/display.rs:
```diff
@@ -418,8 +422,8 @@ impl Display {
-                    padding_x: px,
-                    padding_y: py, ..} = *terminal.size_info();
+                    padding_left: px,
+                    padding_bottom: py, ..} = *terminal.size_info();
```
not really sure about this part. I don't know what this does, but I assume it's a coordinate starting from the left.

chrisduerr commented on src/display.rs:
```diff
@@ -178,8 +180,10 @@ impl Display {
+            padding_top: top as f32,
+            padding_right: right as f32,
+            padding_bottom: bottom as f32,
+            padding_left: left as f32,
```
At this point it would probably be a better idea to store `padding` on the Display rather than having four different fields. Accessing `self.padding.left` should not be any more complicated than `self.padding_left` either.

chrisduerr replied:
This impacts multiple different structs like SizeInfo too, I would propose changing that for most of them if it makes parameters and struct definitions simpler (it should).

chrisduerr commented on src/config.rs:
```diff
@@ -1459,6 +1459,14 @@ pub struct Delta<T: Default> {
+    pub top: u8,
+    pub right: u8,
+    pub bottom: u8,
+    pub left: u8,
```
It might make sense to add `x` and `y` to this struct as a deprecation strategy, so no current configuration files are broken.

sodiumjoe commented:
> What is this useful for except for the window padding at the top because of decorations? If that is the only reason to introduce this, there might be better alternatives.

To be honest, that's pretty much the only reason for this. You're right that there might be better alternatives, like automatically adding an amount of padding for the titlebar for `decorations: transparent`. It seemed simpler to do allow consumers to do it for themselves, though, since as far as I could tell, there isn't a foolproof way to get the titlebar height across different versions of macos.

Personally I don't even plan to use this, since I prefer the `decorations: buttonless` setting, but it came up so many times and seemed easy enough to implement I took a pass.

Totally open to other implementations, we should probably figure that out before I address the rest of your feedback.

chrisduerr commented:
It just seems kinda odd to convolute the config file for a setting that is only used on macos and only used because we don't know how to automatically calculate the offset required.

albertorestifo commented:
What about a CSS-style padding setting, where you can specify the padding in a clockwise order?

- `padding.y: 10` would mean padding of 10px both at the top and bottom of the viewport
- `padding.y: 25 10` would mean padding of 25px on the top and 10px on the bottom.
 
Same goes for `padding.x: <right> <left>`

This avoids adding a breaking change in the configuration, maintains the current behaviour by default but still allows to add more granular padding for the few times where it's useful

chrisduerr commented:
I feel like this would add too much complexity and potential for breakage. I don't think `yml` supports automatically parsing fields like that to like a slice with at least one element, so I'd rather just have simpler fields which are less prone to errors.

Good idea though.

sodiumjoe commented:
@chrisduerr

I added a custom deserializer to `Padding` so that `x` and `y` still work, but you can granularly override them with `top|right|bottom|left`. I think this is a pretty clean and flexible compromise.

I also added a `padding: Padding` field to `SizeInfo` and `ShaderProgram`, I think that's what you meant [here](https://github.com/jwilm/alacritty/pull/1290#discussion_r185562050).

I also added a `.horizontal` and `.vertical` method to `Padding` that makes some of the logic a little easier to read.

I still need to fix tests, but wanted to get some eyes on this first.

chrisduerr commented on src/config.rs:
```diff
@@ -1459,6 +1459,101 @@ pub struct Delta<T: Default> {
+
+    }
+}
+
```
This seems far more complicated than necessary. Why didn't you just add the `x` and `y` field to your current `Padding` struct? Then you could just check if those are set and emit a deprecation warning and update top/right/bottom/left.

sodiumjoe replied:
Oh, I was thinking we could just leave the `x` and `y` fields since most people only care about those anyway. And generally, even if people care about `top`, they could just set:
```
  x: 2
  top: 25
  bottom: 2
```

or whatever. But if we want to just fully migrate over to the new format, I can do that.

chrisduerr replied:
I think the redundanc of having both `y` and `top/bottom` would confuse people. Keeping around `y` as an option seems like a viable path, but I feel like it just creates unnecessary complexity where the user might have `y` + `top/bottom` set and we run into a scenario where the result is not directly obvious (unless explicitly documented).

So for the sake of simplifying the code and having just one way to do things instead of two, I'd rather see a complete migration with deprecation of `x/y` instead of a mix of the two. That's just my personal opinion though.

sodiumjoe replied:
ok, I'll try to get that done tonight

chrisduerr commented on src/display.rs:
```diff
@@ -178,8 +180,7 @@ impl Display {
             cell_height: cell_height as f32,
-            padding_x: config.padding().x as f32,
-            padding_y: config.padding().y as f32,
+            padding: padding.clone(),
```
Cloning the padding seems like a waste. It should be perfectly fine for padding to implement `Copy`.

chrisduerr commented on src/config.rs:
```diff
@@ -269,23 +269,59 @@ pub struct WindowConfig {
+                    padding.left = x;
+                    padding.right = x;
+                        eprintln!("{}", fmt::Yellow("Config `padding.x` is deprecated. Please use \
+                                                    `right|left` instead"));
```
Tiny formatting hick-up here. `eprintln` should be indented properly.

chrisduerr commented on src/config.rs:
```diff
@@ -269,23 +269,59 @@ pub struct WindowConfig {
+    }
+}
+
+fn default_padding_field() -> u8 {
```
I think that setting `x/y` to just `None` would be perfectly fine then there wouldn't be the scenario where `x/y` is removed because of deprecation but the method still stays around (linter should warn though). But I think this is totally fine the way it is too.

chrisduerr replied:
Also there's no reason to have a `default_padding` method. `impl Default for Padding` seems like a better choice.
This would also allow using `..Default::default()` for `x/y`.

sodiumjoe replied:
I was just trying to work with the existing `default_padding` function, but you're right that's much better.

chrisduerr commented on src/config.rs:
```diff
@@ -269,23 +269,59 @@ pub struct WindowConfig {
+                    eprintln!("{}", fmt::Yellow("Config `padding.y` is deprecated. Please use \
+                                                `top|bottom` instead"));
+                },
+                None => {}
```
Instead of doing `match padding.y { Some(y) => { ... }, None => {} }` it would be more ergonomic to use `if let Some(y) => { ... }`. Like this the `None => {}` part which does nothing can just be left out.

sodiumjoe replied:
Ah yes that's much nicer. I first tried with `if padding.y.is_some()`, but that involved an `unwrap`, `if let` was exactly what I was looking for, thanks!

chrisduerr commented on src/config.rs:
```diff
@@ -1459,6 +1495,27 @@ pub struct Delta<T: Default> {
+    x: Option<u8>,
+    #[serde(default="default_padding_field_deprecated")]
+    y: Option<u8>,
+}
```
I think these would force the whole config to the default when someone specifies a `u16` or another invalid value for one of the fields. To make sure that this doesn't happen the `deseralize_with` option in serde should be used similarly to the other places where it's done.

Because the default padding is not 0, you'd probably have to add a custom deserializer that deserializes u8 to `default_padding`, but that shouldn't be too complex (as mentioned there should be multiple examples in the config). So just a simple `deserialize_with="deserialize_padding_side"` (or similar) should do the trick.

sodiumjoe replied:
i'm not entirely sure this is what you meant, but i've attempted to implement your feedback. The only weird thing left here is that the `struct Padding` definition doesn't actually use the `deserialize_padding` function, instead just using `derive(Deserialize)`, which has different behavior, I think. Maybe this doesn't matter, since this doesn't happen outside the context of a `Config` (which is where `deseerialize_padding` is used)?

chrisduerr replied:
Since you have derived Deserialize on the Padding struct, you actually don't need the `deserialize_padding` function anymore. Instead you can make use of `failure_default` serializer to handle the potential deserialization error for you.

This change should allow you to get rid of a big block of code. The derpecation warnings printed inside there could be moved to `deserailize_padding_side` instead.

chrisduerr commented on src/renderer/mod.rs:
```diff
@@ -1017,8 +1019,7 @@ impl ShaderProgram {
             u_background: background,
-            padding_x: config.padding().x,
-            padding_y: config.padding().y,
+            padding: config.padding().clone(),
```
Another clone that shouldn't be necessary, same as the other one, this can be removed by making padding `Copy`.

sodiumjoe replied:
For these, I've changed them to simply dereferencing `config.padding()`, since they're already `derive(Copy)`. This appears to work.

chrisduerr commented:
You could do it similarly to what I've done here: https://github.com/jwilm/alacritty/pull/1279/files#diff-6d8a5e2104954a026de2bfe2b936aea4R1403.

But as mentioned in other issues I *really* don't like the current way alacritty does configuration fallback/deprecation. We should strive to create a general-purpose solution for that rather than "hacking" it into serde which was not made for this purpose. A serde wrapper would probably make this a lot easier.

sodiumjoe commented:
@chrisduerr Ok, I think I've got it now. Your suggestions made a lot of sense and things are much cleaner and rely more on `serde` now. It looks like it'll be pretty easy to remove the deprecation-handling code in the future as well.

> But as mentioned in other issues I really don't like the current way alacritty does configuration fallback/deprecation. We should strive to create a general-purpose solution for that rather than "hacking" it into serde which was not made for this purpose. A serde wrapper would probably make this a lot easier.

I agree, but it seems like handling cases like this, where we're applying deprecated fields as the new fields, for example, would be tricky to generalize over. Granular fallback seems feasible, but deprecation strategies seem like they would depend a lot on the specifics of the field being deprecated, and what the new functionality is.

sodiumjoe commented:
I still need to fix the tests, I'll try to get to that tonight.

chrisduerr commented:
Looks a lot better now, nice! In theory it shouldn't be too hard to "map" a deprecated field to the new value either. An annotation like `#[deprecated="config.padding.y]` should be possible. At least when it's directly from one field to one other field (not in this case where we go from one to two other fields, but still).

sodiumjoe commented:
tests are now fixed

sodiumjoe commented:
rebased master

chrisduerr commented:
I'm still not sure if this PR is the right way to solve the problem with decorations on macOS.

There has been some further discussion in https://github.com/jwilm/alacritty/issues/1671, but without any good solution either.

Potentially using a hardcoded padding at the top, scaled by DPI, might be a solution to this problem which doesn't introduce any changes to the configuration file? I'm not sure how the window controls for maximize/minimize/close behave on the different macOS platforms with different DPI monitors.

If there is any other usecase for this, other than working around the title bar spacing, it might be a good workaround for the time being. But I'm not sure if this is any use at all besides the macOS title bar problem.

sodiumjoe commented:
Yeah, I basically agree with all your points. As I've mentioned before, I wouldn't even use this since I prefer `decorations: buttonless`, so I'd be ok with closing this and letting any interested parties continue the discussion/investigation on #1671.

WeetHet commented:
> I'm still not sure if this PR is the right way to solve the problem with decorations on macOS.
> 
> There has been some further discussion in #1671, but without any good solution either.
> 
> Potentially using a hardcoded padding at the top, scaled by DPI, might be a solution to this problem which doesn't introduce any changes to the configuration file? I'm not sure how the window controls for maximize/minimize/close behave on the different macOS platforms with different DPI monitors.
> 
> If there is any other usecase for this, other than working around the title bar spacing, it might be a good workaround for the time being. But I'm not sure if this is any use at all besides the macOS title bar problem.

Why don't just implement this and leave configuration file as it, having additional top/bottom/left/right besides x and y, so old configurations don't break and new ones can use it for transparent style on macOS

Lambdaphile commented:
@WeetHet Because they're not using this setting... Just WTF...
